### ISSUE TITLE
Patch for u_certificate_provisioning.cpp

### ISSUE THREAD

**Issue by akhilles** [apache/teaclave-sgx-sdk#59]
I apologize if this was explained elsewhere but why is the patch for u_certificate_provisioning.cpp needed? I understand the trts patch.

Is it at all possible to use this SDK with an unmodified aesm service?

Thanks!

**Comment by akhilles** [comment]
Also, can the compiler-rt patch be incorporated into the sdk patch?

EDIT: Actually never mind, the current solution w/ statically linking the patch is probably better.

**Comment by dingelish** [comment]
The first question about the aesm patch:
If not patched, the mutual-ra sample won't work because [ocall_get_update_info](https://github.com/baidu/rust-sgx-sdk/blob/de3b4d21e4a0ef383086c6a22a32ee176492aa23/samplecode/mutual-ra/enclave/src/cert.rs#L330) returns non-zero. This function directly makes an ocall to [sgx_report_attestation_status](https://github.com/baidu/rust-sgx-sdk/blob/de3b4d21e4a0ef383086c6a22a32ee176492aa23/samplecode/mutual-ra/app/src/main.rs#L147) which triggers the certificate check with wrong key endian.

For the compiler-rt patch:
When we first linked everything together without this compiler-rt patch, some error like 'missing ror4' function occurred. Then we found that we can eliminate this error by simply add a '--gc-section' [linker flag](https://github.com/baidu/rust-sgx-sdk/blob/de3b4d21e4a0ef383086c6a22a32ee176492aa23/samplecode/hello-rust/Makefile#L114) to eliminate this bug. We figured out the reason is that the compiler-rt does not provide that function though not needed in the final program. Then we decided to provide a patch in case something else would be needed in the future. So now if you simply remove everything around compiler-rt patch but keep the `--gc-section` flag, everything should be fine. But it is not guaranteed to work in future (I haven't test this project without this patch)

Let me try to do something more about this compiler-rt patch. I think the best way is to embed it to the [compiler-builtins](https://github.com/rust-lang-nursery/compiler-builtins) of Rust.
